### PR TITLE
feat(eventstream): Add delete_tag message types and implement for Snu…

### PR DESCRIPTION
**Pull request by bretthoerner**
…ba tagstore backend

### CODE REVIEW

File: src/sentry/tagstore/snuba/backend.py [hunk 1]
```diff
@@ -689,6 +689,11 @@ def get_group_ids_for_search_filter(
         # instead.
         raise NotImplementedError
 
+    def delete_tag_key(self, project_id, key):
+        from sentry import eventstream
+        eventstream_state = eventstream.start_delete_tag(project_id, key)
+        eventstream.end_delete_tag(eventstream_state)
```
Comment (bretthoerner) [inline src/sentry/tagstore/snuba/backend.py]: We use `start_` and `end_` to:

1. Allow us to do Sentry-side logic in-between the two in the future, as other messages do.
2. Maintain a relatively unified API with the other messages, at very low cost.

File: src/sentry/tagstore/snuba/backend.py [hunk 2]
```diff
@@ -689,6 +689,11 @@ def get_group_ids_for_search_filter(
...
+
 
 class SnubaCompatibilityTagStorage(SnubaTagStorage):
```
Comment (bretthoerner) [inline src/sentry/tagstore/snuba/backend.py]: It's unclear to me whether my `delete_tag_key` impl belongs in `SnubaCompatibilityTagStorage` or in the real one (where it is now).

**Review by tkaemming — CHANGES_REQUESTED** [review]

File: src/sentry/tagstore/snuba/backend.py (hunk 2 above)
Comment (bretthoerner) [reply src/sentry/tagstore/snuba/backend.py]: > I do wonder if the `eventstream` calls should just wrap the call site instead, and we should move the actual call to `delete_tag_key` into the compatibility backend as a noop based on the precedent from the other operations?

Good call.

File: src/sentry/tagstore/snuba/backend.py (hunk 1 above)
Comment (tkaemming) [reply src/sentry/tagstore/snuba/backend.py]: Ah right, duh.

File: src/sentry/tagstore/snuba/backend.py (hunk 2 above)
Comment (bretthoerner) [reply src/sentry/tagstore/snuba/backend.py]: Although -- that works today (and hopefully forever) but we have this weird case like the other endpoints where `tagstore.delete_tag_key` could fire off an async task which itself needs to send the `end` message.

Buuut... we can fix that if we ever come to it? And hopefully we never have to? (Long live Snuba)

Comment (bretthoerner) [reply src/sentry/tagstore/snuba/backend.py]: OK, I moved it to the API call site, so it's not dependent on the backend.

I *did* however punt on passing down the `eventstream_state` and requiring the `tagstore.delete_tag_key` callee to handle sending the end message. This means for the existing v2 backend the `end_delete_tag` message will most likely be sent before the v2 Celery task finishes running.

That seemed like a lot of effort and ugly APIs to perfectly align the timestamp for a message that has nothing to do with v2 in practice.

Comment (tkaemming) [reply src/sentry/tagstore/snuba/backend.py]: > I _did_ however punt on passing down the `eventstream_state` and requiring the `tagstore.delete_tag_key` callee to handle sending the end message. This means for the existing v2 backend the `end_delete_tag` message will most likely be sent before the v2 Celery task finishes running.

I think that's a reasonable compromise, and probably worth noting here as a comment?

File: src/sentry/eventstream/snuba.py [hunk 3]
```diff
@@ -42,3 +42,9 @@ def start_unmerge(self, project_id, hashes, previous_group_id, new_group_id):
 
     def end_unmerge(self, state):
         pass
+
+    def start_delete_tag(self, project_id, tag):
+        pass
+
+    def end_delete_tag(self, state):
+        pass
```
Comment (tkaemming) [inline src/sentry/eventstream/snuba.py]: Is there a plan to expose this via the HTTP API (like `insert_raw` above) for local development?

**Review by tkaemming — APPROVED** [review]
For future reference (I was trying to find it earlier), the other half of this is over at getsentry/snuba#270.

File: src/sentry/eventstream/snuba.py (hunk 3 above)
Comment (bretthoerner) [reply src/sentry/eventstream/snuba.py]: Yeah, we should probably do all of them (delete_group, merge, unmerge, delete_tag). I may take a looksee tomorrow.

Comment (bretthoerner) [reply src/sentry/eventstream/snuba.py]: https://github.com/getsentry/snuba/pull/274

**Pull request merged**